File: stampli-backend/src/main/java/com/example/modules/authentication/service/register/EventedRegisterService.java
package com.example.modules.authentication.service.register;

import org.springframework.context.ApplicationEventPublisher;

import javax.transaction.Transactional;


public record EventedRegisterService(RegisterService registerService,
                                     ApplicationEventPublisher eventPublisher) implements RegisterService {

    @Transactional
    public Integer registerUser(String email, String password) {
        var userId = registerService.registerUser(email, password);

        eventPublisher.publishEvent(new UserRegisteredEvent(userId));

        return userId;
    }

    public Integer registerUserWithToken(String issuer, String id) {
        var userId = registerService.registerUserWithToken(issuer, id);

        eventPublisher.publishEvent(new UserRegisteredEvent(userId));

        return userId;
    }
}

File: stampli-backend/src/main/java/com/example/modules/authentication/service/register/EventedRegisterService.kt
package com.example.modules.authentication.service.register

import org.springframework.context.ApplicationEventPublisher
import javax.transaction.Transactional

open class EventedRegisterService(val registerService: RegisterService, val eventPublisher: ApplicationEventPublisher) :
    RegisterService by registerService {

    @Transactional //
    override fun registerUser(email: String?, password: String?): Int {
        val userId = registerService.registerUser(email, password)

        eventPublisher.publishEvent(UserRegisteredEvent(userId));

        return userId;
    }

    override fun registerUserWithToken(issuer: String?, id: String?): Int {
        val userId = registerService.registerUserWithToken(issuer, id)

        eventPublisher.publishEvent(UserRegisteredEvent(userId));

        return userId;
    }
}
